docs/using-the-compiler.rst

```diff
@@ -0,0 +1,249 @@
+          }
+      ]
+      contracts: {
+        "sourceFile.sol:ContractName": {
+          // The Ethereum Contract ABI. If empty, it is represented as an empty array.
+          // See https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI
+          abi: [],
+          evm: {
+              assembly:
+              bytecode:
+              runtimeBytecode:
+              opcodes:
+              annotatedOpcodes: // (axic) see https://github.com/ethereum/solidity/issues/1178
+              gasEstimates:
+              sourceMap:
+              runtimeSourceMap:
+              // If given, this is an unlinked object (cannot be filtered out explicitly, might be
+              // filtered if both bytecode, runtimeBytecode, opcodes and others are filtered out)
+              linkReferences: {
+                "sourceFile.sol:Library1": [1, 200, 80] // byte offsets into bytecode. Linking replaces the 20 bytes there.
+              }
+              // the same for runtimeBytecode - I'm not sure it is a good idea to allow to link libraries differently for the runtime bytecode.
+              // furthermore, runtime bytecode is always a substring of the bytecode anyway.
+              runtimeLinkReferences: {
+              }
+          },
+          functionHashes:
+          metadata: // see below
+          ewasm: {
+              wast: // S-expression format
+              wasm: //
+          },
+          userdoc: // Obsolete
+          devdoc: // Obsolete
+          natspec: // Combined dev+userdoc
+        }
+      },
+      formal: {
+        "why3": "..."
+      },
+      sourceList: ["source1.sol", "source2.sol"], // this is important for source references both in the ast as well as in the srcmap in the contract
+      sources: {
+        "source1.sol": {
+          "AST": { ... }
+        }
+      }
+    }
+
+Metadata Output
+---------------
+
+Note that the actual bytecode is not part of the metadata because the hash
+of the metadata structure will be included in the bytecode itself.
+
+This requires the compiler to be able to compute the hash of its own binary,
+which requires it to be statically linked. The hash of the binary is not
+too important. It is much more important to have the commit hash because
+that can be used to query a location of the binary (and whether the version is
+"official") at a registry contract.
+
+    {
+      version: "1",
+      language: "Solidity",
+      compiler: {
+        commit: "55db20e32c97098d13230ab7500758e8e3b31d64",
+        version: "soljson-2313-2016-12-12",
+        keccak: "0x123..."
+      },
+      // This is a subset of the regular compiler input
+      sources:
+      {
+        "abc": {keccak: "0x456..."}, // here, sources are always given by hash
+        "def": {keccak: "0x123..."},
+        "dir/file.sol": {keccax: "0xabc..."},
+        "xkcd": {swarm: "0x456..."}
+      },
+      // This is a subset of the regular compiler input
```

Comment:
`settings` is up to the intrepretation of the compiler defined by the `language` and `compiler` fields.